Fix Socks5() connect failures to be less noisy and less unnecessarily scary

wtogami opened the pull request:
Make failures to connect via Socks5() more informative and less unnecessarily scary.
- The "ERROR" was printed far too often during normal operation for what was not an error.
- Makes the Socks5() connect failure similar to the IP connect failure in debug.log.

Before:
`2016-05-09 00:15:00 ERROR: Proxy error: host unreachable`

After:
`2016-05-09 00:15:00 Socks5() connect to t6xj6wilh4ytvcs7.onion:18333 failed: host unreachable"`

SOCKS5 connecting and connected messages with -debug=net.

They were too noisy and not necessary for normal operation.

wtogami committed:
Make failures to connect via Socks5() more informative and less unnecessarily scary.

* The "ERROR" was printed far too often during normal operation for what was not an error.
* Makes the Socks5() connect failure similar to the IP connect failure in debug.log.

Before:
`2016-05-09 00:15:00 ERROR: Proxy error: host unreachable`

After:
`2016-05-09 00:15:00 Socks5() connect to t6xj6wilh4ytvcs7.onion:18333 failed: host unreachable"`

wtogami committed:
SOCKS5 connecting and connected messages with -debug=net.

They were too noisy and not necessary for normal operation.

pstratem commented:
utACK 0d9af79e5084dc2fb6a73abd9dd113dda4e993b4

We need another PR to honor logips=0

wtogami commented:
Honoring logips=0 is beyond the scope of this PR, various other things need to be fixed too.

laanwj commented:
ping @cfields

laanwj commented:
ping @theuni

midnightmagic commented:
yes plz! :-)

laanwj commented on src/netbase.cpp:
```diff
@@ -379,19 +379,21 @@ static bool Socks5(const std::string& strDest, int port, const ProxyCredentials
-            case 0x07: return error("Proxy error: protocol error");
-            case 0x08: return error("Proxy error: address type not supported");
-            default:   return error("Proxy error: unknown");
+            case 0x01: LogPrintf("Socks5() connect to %s:%d failed: general failure\n", strDest, port); break;
```
I'd prefer assigning the error to a string here (or make a `string Socks5ErrorString(code)` function), instead of repeating the `Socks5() connect to %s:%d failed` on every line.

wtogami replied:
@phantomcircuit informs me that other non-error cases here print scary "ERROR:" messages outside of `if (pchRet2[1] != 0x00)`, so it is not straight forward to make a function that decodes only the Socks5ErrorString().  I originally thought about adding a function similar to error() but then I was told all of this needs to be rewritten for the network rewrite so don't worry about it for now.

theuni replied:
Yes, this has been rewritten anyway. Adding to what @laanwj said, it'd be useful to have the code in the return so that the caller could react appropriately (for ex "address type not supported" should be a hint not to try that type again). In the refactored code, this returns a general type error (connection failed/proxy failed/etc), as well as a specific failure type.

laanwj replied:
It's not like I'm asking for much, you can literally take this over:
```c++
std::string Socks5ErorString(int err)
{
    switch(err) {
        case 0x01: return "general failure";
        case 0x02: return "connection not allowed";
        case 0x03: return "network unreachable";
        case 0x04: return "host unreachable";
        case 0x05: return "connection refused";
        case 0x06: return "TTL expired";
        case 0x07: return "protocol error";
        case 0x08: return "address type not supported";
        default:   return "unknown";
    }
}
```
Then replace the switch-and-LogPrintf here with:
```c++
LogPrintf("Socks5() connect to %s:%d failed: %s\n", strDest, port, Socks5ErrorString(pchRet2[1]));
```

theuni commented:
No strong feelings either way about the errors. Switching the debug stuff to "net" is obviously a good change, though.

pstratem commented:
This is certainly not perfect, but is am improvement.

utACK 0d9af79e5084dc2fb6a73abd9dd113dda4e993b4

wtogami committed:
Make Socks5() InterruptibleRecv() timeout/failures informative.

Before:
2016-05-16 06:10:45 ERROR: Error reading proxy response

After:
2016-05-16 06:10:45 Socks5() connect to k7s5d6jqig4ej4v4.onion:18333 failed: InterruptibleRecv() timeout or other failure

wtogami commented:
94fd1d8 makes the last remaining non-ERROR more informative and less scary.

MarcoFalke commented:
utACK 94fd1d8

wtogami commented:
I believe this is ready for merge.

wtogami committed:
Use Socks5ErrorString() to decode error responses from socks proxy.

wtogami commented:
Followed @laanwj suggestion to add Socks5ErrorString() to decode error responses from socks proxy.

Note the separate LogPrintf in 94fd1d8d53adceb80e5a41cc6438c7704aeac0f7 is outside of what Socks5ErrorString() can decode.

Is this sufficient, do you want a squash?

laanwj commented:
Looks good to me, thanks.
Yes the case in  https://github.com/bitcoin/bitcoin/commit/94fd1d8d53adceb80e5a41cc6438c7704aeac0f7 is separate, that doesn't matter, it's not an error returned from the proxy server but a local network error.
